ivadomed/loader/mri2d_segmentation_dataset.py

```diff
@@ -147,17 +147,28 @@ def prepare_indices(self):
                 if length > size:
                     raise RuntimeError('"length_2D" must be smaller or equal to image dimensions after resampling.')
 
-            for x in range(0, (shape[0] - self.length[0] + self.stride[0]), self.stride[0]):
-                if x + self.length[0] > shape[0]:
-                    x = (shape[0] - self.length[0])
-                for y in range(0, (shape[1] - self.length[1] + self.stride[1]), self.stride[1]):
-                    if y + self.length[1] > shape[1]:
-                        y = (shape[1] - self.length[1])
+            for x_min in range(0, (shape[0] - self.length[0] + self.stride[0]), self.stride[0]):
+                if x_min + self.length[0] > shape[0]:
+                    x_min = (shape[0] - self.length[0])
+                x_max = x_min + self.length[0]
+                for y_min in range(0, (shape[1] - self.length[1] + self.stride[1]), self.stride[1]):
+                    if y_min + self.length[1] > shape[1]:
+                        y_min = (shape[1] - self.length[1])
+                    y_max = y_min + self.length[1]
+
+                    patch = {'input': list(np.asarray(self.handlers[i][0]['input'])[:, x_min:x_max, y_min:y_max]),
+                             'gt': list(np.asarray(self.handlers[i][0]['gt'])[:, x_min:x_max, y_min:y_max]) \
+                                   if self.handlers[i][0]['gt'] else [],
+                             'input_metadata': self.handlers[i][0]['input_metadata'],
+                             'gt_metadata': self.handlers[i][0]['gt_metadata']}
+                    if self.slice_filter_fn and not self.slice_filter_fn(patch, is_2d_patch=self.is_2d_patch):
```

Comment:
I am not sure sure what `self.handlers[i][0]` is? Is it a dict, dataframe or something else? I checked the earlier comment on handler and it seems that it is just a list of indexes? So... where did this stem from? My guess so far is that the respective input/gt/keys are the actual image/slice/(source) shape? 

I feel we can afford to be a lot more verbose here to expand it out for ease of understanding as we are packing a lot of actions into a few lines.

Reply:
Yup. Good point. I strongly agree this is not something that has to be changed per this PR since you are just carrying out the earlier design patterns. I was just curious. Thanks for the explanation.

Reply:
Ok thanks, I'll keep that in mind while re-designing the PR

Reply:
As discussed, I added this line for now:
https://github.com/ivadomed/ivadomed/blob/29f24072c6d6b374f7c23292881cf1a44c10ae00/ivadomed/loader/mri2d_segmentation_dataset.py#L163
Let's move the indexes vs. handlers documentation to another issue/PR.

Reply:
Agreed. Thanks for the clarification. I will open an internal refactoring ticket for this.